src/rdkafka_cgrp.c

```diff
@@ -2434,6 +2435,280 @@ static void rd_kafka_cgrp_offset_commit_tmr_cb (rd_kafka_timers_t *rkts,
+                        partitions->cnt, cur_assignment_cnt);
+        }
+
+        /* Verify partitions exist in current assignment before making
+         * any changes to it to ensure the operation doesn't partially
+         * succeed. */
+        for (i = 0; i < partitions->cnt; i++) {
+                const rd_kafka_topic_partition_t *rktpar =
+                        &partitions->elems[i];
+                if (!rd_kafka_topic_partition_list_find(rkcg->rkcg_assignment,
+                                                        rktpar->topic,
+                                                        rktpar->partition)) {
+                        return rd_kafka_error_new(
+                                RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                "%s [%"PRId32"] is not part of the current "
+                                "assignment",
+                                rktpar->topic, rktpar->partition);
+                }
+        }
+
+        /* Whether or not it was before, current assignment is now not lost. */
+        rd_atomic32_set(&rkcg->rkcg_assignment_lost, rd_false);
+
+        if (partitions->cnt == 0) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "UNASSIGN",
+                     "Group \"%s\": removing 0 of %d assigned partition(s) "
+                     "in join state %s (nothing to do)",
+                     rkcg->rkcg_group_id->str, rkcg->rkcg_assignment->cnt,
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+                return NULL;
+        }
+
+        rd_kafka_cgrp_version_new_barrier(rkcg);
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "UNASSIGN",
+                     "Group \"%s\": removing %d of %d assigned partition(s) "
+                     "in join state %s (v%"PRId32")",
+                     rkcg->rkcg_group_id->str, partitions->cnt,
+                     rkcg->rkcg_assignment->cnt,
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state],
+                     rkcg->rkcg_version);
+
+        rd_kafka_cgrp_set_join_state(rkcg,
+                RD_KAFKA_CGRP_JOIN_STATE_WAIT_INCR_UNASSIGN);
+
+        rkcg->rkcg_flags &= ~RD_KAFKA_CGRP_F_WAIT_UNASSIGN;
+
+        /* Remove partitions from the current assignment */
+        for (i = 0; i < partitions->cnt; i++) {
+                const rd_kafka_topic_partition_t *rktpar =
+                        &partitions->elems[i];
+
+                if (!rd_kafka_topic_partition_list_del(rkcg->rkcg_assignment,
+                                                       rktpar->topic,
+                                                       rktpar->partition))
+                        rd_assert(
+                                !*"rd_kafka_topic_partition_list_del failed");
+        }
+
+        /* Update statistics */
+        rd_kafka_wrlock(rkcg->rkcg_rk);
+        rkcg->rkcg_c.assignment_size -= partitions->cnt;
```

Comment:
` = rkcg->rkcg_assignment_cnt` is less error prone.

Reply:
actually more error prone to change given it's there, because there's a chance i'll accidentally leave the +=.